src/engine/mod.rs

```diff
@@ -0,0 +1,657 @@
+//! This module provides the Warpgrapher engine, including supporting modules for
+//! configuration, GraphQL schema generation, resolvers, and interface to the
+//! database.
+
+use actix_web::web::Json;
+
+use super::error::{Error, ErrorKind};
+use config::{WarpgrapherConfig, WarpgrapherProp, WarpgrapherResolvers, WarpgrapherValidators};
+use context::{GraphQLContext, WarpgrapherRequestContext};
+use extensions::WarpgrapherExtensions;
+use juniper::http::GraphQLRequest;
+use log::{debug, error};
+use r2d2::Pool;
+use r2d2_cypher::CypherConnectionManager;
+use schema::{create_root_node, RootRef};
+use serde_json;
+use serde_json::json;
+use std::collections::HashMap;
+use std::env::var_os;
+use std::fmt::Debug;
+use std::option::Option;
+
+pub mod config;
+pub mod context;
+pub mod extensions;
+pub mod neo4j;
+pub mod objects;
+mod resolvers;
+pub mod schema;
+mod visitors;
+
+#[allow(clippy::borrowed_box)]
+fn graphql_error(err: &Box<dyn std::error::Error + Send + Sync>) -> Error {
+    match serde_json::to_string(&json!({ "message": format!("{}", err) })) {
+        Ok(s) => Error::new(ErrorKind::GraphQLError(s), None),
+        Err(e) => {
+            error!("Failed to serialize error object:  {:#?}", e);
+            Error::new(
+                ErrorKind::GraphQLError("INTERNAL SERVER ERROR".to_string()),
+                None,
+            )
+        }
+    }
+}
+
+impl WarpgrapherRequestContext for () {
+    fn new() {}
+}
+
+/// A Warpgrapher GraphQL engine.
+///
+/// The [`Engine`] struct Juniper GraphQL service
+/// on top of it, with an auto-generated set of resolvers that cover basic CRUD
+/// operations, and potentially custom resolvers, on a set of data types and
+/// the relationships between them.  The engine includes handling of back-end
+/// communications with the chosen databse.
+///
+/// [`Engine`]: ./struct.Engine.html
+///
+/// # Examples
+///
+/// ```rust
+/// use warpgrapher::{Engine, Neo4jEndpoint};
+/// use warpgrapher::engine::config::WarpgrapherConfig;
+///
+/// let config = WarpgrapherConfig::default();
+/// let db = Neo4jEndpoint::from_env("DB_URL").unwrap();
+///
+/// let mut engine = Engine::<(), ()>::new(config, db)
+///     .build().unwrap();
+///
+/// ```
+#[derive(Clone)]
+pub struct Engine<GlobalCtx = (), ReqCtx = ()>
+where
+    GlobalCtx: 'static + Clone + Sync + Send + Debug,
+    ReqCtx: 'static + Clone + Sync + Send + Debug + WarpgrapherRequestContext,
+{
+    pub config: WarpgrapherConfig,
+    pub database: String,
+    pub pool: Option<Pool<CypherConnectionManager>>,
```

Comment:
Does it ever make sense to construct an engine without a database pool?  If not, this probably shouldn't be an option.  If so, then fine as is.

Reply:
This and the root_node value get constructed in the build() operation since they are not direct pass-through values (they are derived from the database string and the config).

Reply:
To address this, will implement an EngineBuilder